### PR TITLE
[breaking] Implement the es-shim-api interface

### LINKED ISSUE #1: version that doesn't punch globals
This is really useful code but it punches the global `String.prototype.codePointAt`, which doesn't compose well (thrashing/breakage likely when multiple competing modules try to "fix" the same thing). Can we get a version of this that just does a module.exports= instead of mutating a global?

### PR DESCRIPTION
**Pull request by nicolo-ribaudo**
Fixes #1 

@ljharb Am I doing it correctly?

---

@mathiasbynens Here is a summary of the changes in this PR. Basically, I rewrote all the infrastructure while keeping the original implementation.

1. I changed the package so that it follows the `es-shim-api` interface. Since it requires multiple files, it's now harder to directly use this package in the browser. However, there is an easy (even if not recommended in production) workaround using `bundle.run`.
2. Instead of inlining the abstract operations and the `defineProperty` IE check, I used the `es-abstract` and `define-properties` packages. This has the disadvantage of a new dependency, but they have 20M and 13M dowloads/month on npm so they are likely already in the `node_modules` of our users.
3. I changed the tests to provide a complete output in case of multiple failures, and added every node version to travis.

Is this ok?

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -3,6 +3,7 @@ coverage
 
 # Installed npm modules
 node_modules
+package-lock.json
```
Comment (ljharb) [inline .gitignore:6]: you’ll want to add `package-lock=false` to `.npmrc` as well

File: index.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
+/*! https://mths.be/codepointat v1.0.0 by @mathias */
+
```
Comment (ljharb) [inline index.js:2]:
```suggestion

'use strict;'

```

File: package.json [hunk 3]
```diff
@@ -3,7 +3,7 @@
 	"version": "0.2.1",
 	"description": "A robust & optimized `String.prototype.codePointAt` polyfill, based on the ECMAScript 6 specification.",
 	"homepage": "https://mths.be/codepointat",
-	"main": "codepointat.js",
+	"main": "index.js",
```
Comment (ljharb) [inline package.json:6]:
```suggestion
	"main": "index.js",
	"exports": {
		".": "./index.js",
		"./auto": "./auto.js",
		"./shim": "./shim.js",
		"./getPolyfill": "./getPolyfill.js",
		"./implementation": "./implementation.js",
		"./package.json": "./package.json"
	},
```

File: package.json [hunk 4]
```diff
@@ -21,12 +21,17 @@
 		"url": "https://github.com/mathiasbynens/String.prototype.codePointAt.git"
 	},
 	"bugs": "https://github.com/mathiasbynens/String.prototype.codePointAt/issues",
-	"files": [
-		"LICENSE-MIT.txt",
-		"codepointat.js"
-	],
 	"scripts": {
-		"test": "node tests/tests.js",
+		"test": "es-shim-api --bound && npm run test-only",
```
Comment (ljharb) [inline package.json]:
```suggestion
		"pretest": "es-shim-api --bound",
		"test": "npm run test-only",
```

File: shim.js [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+
+module.exports = function shimCodePointAt() {
+    var polyfill = getPolyfill();
+
+    if (String.prototype.codePointAt !== polyfill) {
+        define(String.prototype, { codePointAt: polyfill });
+    }
+
+    return polyfill;
+}
```
Comment (ljharb) [inline shim.js]: trailing newline

**Review by ljharb — CHANGES_REQUESTED** [review]
Looks good overall; I’d probably restructure the tests to match how i do it in the other es-shim packages tho.

File: .travis.yml [hunk 6]
```diff
@@ -2,4 +2,4 @@ language: node_js
 node_js:
   - "0.10"
 script:
-  "node tests/tests.js"
+  "npm i && npm test"
```
Comment (ljharb) [inline .travis.yml]:
```suggestion
  "npm test"
```
`npm install` is already done by default in travis on the "install" step, before "script"

File: .travis.yml [hunk 7]
```diff
@@ -2,4 +2,4 @@ language: node_js
 node_js:
```
Comment (ljharb) [inline .travis.yml]: you might want to run this on a lot more node versions; and might want the linter to run separately from tests - if so, you can follow this pattern: https://github.com/ljharb/unbox-primitive/blob/master/.travis.yml#L1-L7

the only hitch is you'd need to define `"test"` as `"npm run tests-only"`, and put the actual test command in `"tests-only"`.

Comment (nicolo-ribaudo) [reply .travis.yml]: Nice repository!

**Comment by nicolo-ribaudo** [comment]
Oh, Travis isn't running 🤔

**Review by mathiasbynens — APPROVED** [review]

File: package.json [hunk 8]
```diff
@@ -31,7 +31,8 @@
 	"bugs": "https://github.com/mathiasbynens/String.prototype.codePointAt/issues",
 	"scripts": {
 		"pretest": "es-shim-api --bound",
-		"test": "tape 'tests/*.js'",
+        "test": "npm run test-only",
```
Comment (ljharb) [inline package.json]:
```suggestion
		"test": "npm run test-only",
```

**Review by ljharb — APPROVED** [review]

**Comment by nicolo-ribaudo** [comment]
Travis: https://travis-ci.org/github/mathiasbynens/String.prototype.codePointAt/builds/688006369

**Comment by nicolo-ribaudo** [comment]
Travis is red just because I cancelled a redundant build.

**Pull request merged**